# Subject: Classical Data Analysis

## Session 2 - Regression with multiple variables

### Exercise 1 -  Linear Regression with multiple variables 

Considering the Stepwise Regression presented in Demo 1 perform the Linear Regression with multiple variable considering the following independent variables:
    
- Situation 1: radio and newspaper;
- Situation 2: TV, radio and newspaper.
    
Interpret and discuss the OLS Regression Results, namely the different R-squared (R2). 

Commit scripts in your GitHub account. You should export your solution code (.ipynb notebook) and push it to your repository “ClassicalDataAnalysis”.

The following are the tasks that should complete and synchronize with your repository “ClassicalDataAnalysis” until November 10. Please notice that none of these tasks is graded, however it’s important that you correctly understand and complete them in order to be sure that you won’t have problems with further assignments.

# 1. Data

Let's take a look again at some data - Advertising - , ask some questions about that data, and then use linear regression (with one and multiple variables) to answer those questions!

In [1]:
# imports
import pandas as pd
import statsmodels.api as sm
import matplotlib.pyplot as plt
%matplotlib inline

/Users/sergi/anaconda/lib/python3.6/site-packages/statsmodels/compat/pandas.py:56: FutureWarning: The pandas.core.datetools module is deprecated and will be removed in a future version. Please use the pandas.tseries module instead.
  from pandas.core import datetools


In [2]:
# read data into a DataFrame
df = pd.read_csv('http://www-bcf.usc.edu/~gareth/ISL/Advertising.csv', index_col=0)
df.head()

TV  radio  newspaper  sales
1  230.1   37.8       69.2   22.1
2   44.5   39.3       45.1   10.4
3   17.2   45.9       69.3    9.3
4  151.5   41.3       58.5   18.5
5  180.8   10.8       58.4   12.9

In [3]:
df

TV  radio  newspaper  sales
1    230.1   37.8       69.2   22.1
2     44.5   39.3       45.1   10.4
3     17.2   45.9       69.3    9.3
4    151.5   41.3       58.5   18.5
5    180.8   10.8       58.4   12.9
6      8.7   48.9       75.0    7.2
7     57.5   32.8       23.5   11.8
8    120.2   19.6       11.6   13.2
9      8.6    2.1        1.0    4.8
10   199.8    2.6       21.2   10.6
11    66.1    5.8       24.2    8.6
12   214.7   24.0        4.0   17.4
13    23.8   35.1       65.9    9.2
14    97.5    7.6        7.2    9.7
15   204.1   32.9       46.0   19.0
16   195.4   47.7       52.9   22.4
17    67.8   36.6      114.0   12.5
18   281.4   39.6       55.8   24.4
19    69.2   20.5       18.3   11.3
20   147.3   23.9       19.1   14.6
21   218.4   27.7       53.4   18.0
22   237.4    5.1       23.5   12.5
23    13.2   15.9       49.6    5.6
24   228.3   16.9       26.2   15.5
25    62.3   12.6       18.3    9.7
26   262.9    3.5       19.5   12.0
27   142.9   29.3       12.6   15.0
28   240.1   16.7       22.9   15.9
29   248.8   27.1       22.9   18.9
30    70.6   16.0       40.8   10.5
..     ...    ...        ...    ...
171   50.0   11.6       18.4    8.4
172  164.5   20.9       47.4   14.5
173   19.6   20.1       17.0    7.6
174  168.4    7.1       12.8   11.7
175  222.4    3.4       13.1   11.5
176  276.9   48.9       41.8   27.0
177  248.4   30.2       20.3   20.2
178  170.2    7.8       35.2   11.7
179  276.7    2.3       23.7   11.8
180  165.6   10.0       17.6   12.6
181  156.6    2.6        8.3   10.5
182  218.5    5.4       27.4   12.2
183   56.2    5.7       29.7    8.7
184  287.6   43.0       71.8   26.2
185  253.8   21.3       30.0   17.6
186  205.0   45.1       19.6   22.6
187  139.5    2.1       26.6   10.3
188  191.1   28.7       18.2   17.3
189  286.0   13.9        3.7   15.9
190   18.7   12.1       23.4    6.7
191   39.5   41.1        5.8   10.8
192   75.5   10.8        6.0    9.9
193   17.2    4.1       31.6    5.9
194  166.8   42.0        3.6   19.6
195  149.7   35.6        6.0   17.3
196   38.2    3.7       13.8    7.6
197   94.2    4.9        8.1    9.7
198  177.0    9.3        6.4   12.8
199  283.6   42.0       66.2   25.5
200  232.1    8.6        8.7   13.4

[200 rows x 4 columns]

# 2 - Linear Regression with multiple variables in Statsmodels

## 2.1. Regression model with multiple variables and with a constant for Situation 1

> Put your code here

In [4]:
X = df[["TV", "radio"]]
y = df["sales"]

In [5]:
X = sm.add_constant(X) 
X.head()

const     TV  radio
1    1.0  230.1   37.8
2    1.0   44.5   39.3
3    1.0   17.2   45.9
4    1.0  151.5   41.3
5    1.0  180.8   10.8

In [6]:
model = sm.OLS(y, X).fit() ## sm.OLS(output, input)

In [7]:
predictions = model.predict(X) 
predictions

1      20.555465
2      12.345362
3      12.337018
4      17.617116
5      13.223908
6      12.512084
7      11.718212
8      12.105516
9       3.709379
10     12.551697
11      7.035860
12     17.256520
13     10.608662
14      8.810951
15     18.444668
16     20.828915
17     12.903865
18     23.241076
19      9.941215
20     14.153846
21     18.121392
22     14.742064
23      6.514172
24     16.544027
25      8.140352
26     15.608021
27     14.967694
28     17.046335
29     19.399541
30      9.159297
         ...    
171     7.389574
172    14.376846
173     7.596578
174    11.960970
175    13.736151
176    24.783526
177    19.964022
178    12.174924
179    16.013844
180    12.378040
181    10.575089
182    13.933696
183     6.564088
184    24.163936
185    18.537949
186    20.779377
187     9.698684
188    17.060279
189    18.620097
190     6.051445
191    12.454978
192     8.405926
193     4.478859
194    18.448761
195    16.463190
196     5.364512
197     8.152375
198    12.7680

In [8]:
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sales   R-squared:                       0.897
Model:                            OLS   Adj. R-squared:                  0.896
Method:                 Least Squares   F-statistic:                     859.6
Date:                Sun, 03 Dec 2017   Prob (F-statistic):           4.83e-98
Time:                        22:57:04   Log-Likelihood:                -386.20
No. Observations:                 200   AIC:                             778.4
Df Residuals:                     197   BIC:                             788.3
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.9211      0.294      9.919      0.000       2.340       3.502
TV             0.0458      0.001     32.909      0.000       0.043       0.048
radio          0.1880      0.008     23.382      0.000       0.172       0.204
==============================================================================
Omnibus:                       60.022   Durbin-Watson:                   2.081
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              148.679
Skew:                          -1.323   Prob(JB):                     5.19e-33
Kurtosis:                       6.292   Cond. No.                         425.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

## 2.2. Regression model with multiple variables and with a constant for Situation 2

> Put your code here

In [9]:
X = df[["TV", "radio", "newspaper"]]
y = df["sales"]

In [10]:
X = sm.add_constant(X) 
X.head()

const     TV  radio  newspaper
1    1.0  230.1   37.8       69.2
2    1.0   44.5   39.3       45.1
3    1.0   17.2   45.9       69.3
4    1.0  151.5   41.3       58.5
5    1.0  180.8   10.8       58.4

In [11]:
model = sm.OLS(y, X).fit() ## sm.OLS(output, input)

In [12]:
predictions = model.predict(X) 
predictions

1      20.523974
2      12.337855
3      12.307671
4      17.597830
5      13.188672
6      12.478348
7      11.729760
8      12.122953
9       3.727341
10     12.550849
11      7.032299
12     17.285129
13     10.577121
14      8.826300
15     18.434366
16     20.819300
17     12.823657
18     23.224957
19      9.951682
20     14.166073
21     18.100767
22     14.740538
23      6.489150
24     16.545933
25      8.146519
26     15.610039
27     14.989514
28     17.051673
29     19.410538
30      9.144024
         ...    
171     7.394980
172    14.358274
173     7.607692
174    11.970939
175    13.744357
176    24.786870
177    19.979373
178    12.162046
179    16.010997
180    12.384555
181    10.587200
182    13.928099
183     6.554670
184    24.133100
185    18.538521
186    20.803011
187     9.691373
188    17.076442
189    18.644306
190     6.051624
191    12.489159
192     8.424019
193     4.466230
194    18.486958
195    16.495300
196     5.370342
197     8.165312
198    12.7859

In [13]:
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sales   R-squared:                       0.897
Model:                            OLS   Adj. R-squared:                  0.896
Method:                 Least Squares   F-statistic:                     570.3
Date:                Sun, 03 Dec 2017   Prob (F-statistic):           1.58e-96
Time:                        22:57:25   Log-Likelihood:                -386.18
No. Observations:                 200   AIC:                             780.4
Df Residuals:                     196   BIC:                             793.6
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.9389      0.312      9.422      0.000       2.324       3.554
TV             0.0458      0.001     32.809      0.000       0.043       0.049
radio          0.1885      0.009     21.893      0.000       0.172       0.206
newspaper     -0.0010      0.006     -0.177      0.860      -0.013       0.011
==============================================================================
Omnibus:                       60.414   Durbin-Watson:                   2.084
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              151.241
Skew:                          -1.327   Prob(JB):                     1.44e-33
Kurtosis:                       6.332   Cond. No.                         454.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

## 2.3. Regression model with multiple variables and without a constant for Situation 1

> Put your code here

In [21]:
X = df[["TV", "radio"]]
y = df["sales"]

In [22]:
X = sm.add_constant(X) 
X.head()

const     TV  radio
1    1.0  230.1   37.8
2    1.0   44.5   39.3
3    1.0   17.2   45.9
4    1.0  151.5   41.3
5    1.0  180.8   10.8

In [23]:
X["const"] = X["const"].apply(lambda x: x*0)
X

const     TV  radio
1      0.0  230.1   37.8
2      0.0   44.5   39.3
3      0.0   17.2   45.9
4      0.0  151.5   41.3
5      0.0  180.8   10.8
6      0.0    8.7   48.9
7      0.0   57.5   32.8
8      0.0  120.2   19.6
9      0.0    8.6    2.1
10     0.0  199.8    2.6
11     0.0   66.1    5.8
12     0.0  214.7   24.0
13     0.0   23.8   35.1
14     0.0   97.5    7.6
15     0.0  204.1   32.9
16     0.0  195.4   47.7
17     0.0   67.8   36.6
18     0.0  281.4   39.6
19     0.0   69.2   20.5
20     0.0  147.3   23.9
21     0.0  218.4   27.7
22     0.0  237.4    5.1
23     0.0   13.2   15.9
24     0.0  228.3   16.9
25     0.0   62.3   12.6
26     0.0  262.9    3.5
27     0.0  142.9   29.3
28     0.0  240.1   16.7
29     0.0  248.8   27.1
30     0.0   70.6   16.0
..     ...    ...    ...
171    0.0   50.0   11.6
172    0.0  164.5   20.9
173    0.0   19.6   20.1
174    0.0  168.4    7.1
175    0.0  222.4    3.4
176    0.0  276.9   48.9
177    0.0  248.4   30.2
178    0.0  170.2    7.8
179    0.0  276.7    2.3
180    0.0  165.6   10.0
181    0.0  156.6    2.6
182    0.0  218.5    5.4
183    0.0   56.2    5.7
184    0.0  287.6   43.0
185    0.0  253.8   21.3
186    0.0  205.0   45.1
187    0.0  139.5    2.1
188    0.0  191.1   28.7
189    0.0  286.0   13.9
190    0.0   18.7   12.1
191    0.0   39.5   41.1
192    0.0   75.5   10.8
193    0.0   17.2    4.1
194    0.0  166.8   42.0
195    0.0  149.7   35.6
196    0.0   38.2    3.7
197    0.0   94.2    4.9
198    0.0  177.0    9.3
199    0.0  283.6   42.0
200    0.0  232.1    8.6

[200 rows x 3 columns]

In [24]:
model = sm.OLS(y, X).fit() ## sm.OLS(output, input)

In [25]:
predictions = model.predict(X) 
predictions

1      21.526871
2      11.700334
3      11.757999
4      18.040390
5      12.461292
6      11.998620
7      10.881890
8      11.210836
9       0.966486
10     11.571401
11      4.992063
12     17.430750
13      9.575390
14      7.138420
15     18.946300
16     21.956174
17     12.342160
18     24.764721
19      8.625593
20     13.710373
21     18.505456
22     14.222750
23      4.470239
24     16.503848
25      6.385799
26     15.244417
27     14.741343
28     17.103943
29     20.031495
30      7.642127
         ...    
171     5.475546
172    13.946937
173     5.810843
174    10.909400
175    12.999476
176    26.709094
177    20.739953
178    11.173057
179    15.718598
180    11.439099
181     9.201929
182    13.256789
183     4.425497
184    25.905865
185    18.939189
186    21.870132
187     8.146207
188    17.243693
189    18.961793
190     3.876581
191    11.850191
192     6.685703
193     1.909411
194    19.044506
195    16.598673
196     2.966993
197     6.321266
198    11.8994

In [26]:
model.summary()

/Users/sergi/anaconda/lib/python3.6/site-packages/statsmodels/regression/linear_model.py:1471: RuntimeWarning: divide by zero encountered in double_scalars
  return np.sqrt(eigvals[0]/eigvals[-1])
/Users/sergi/anaconda/lib/python3.6/site-packages/statsmodels/base/model.py:1036: RuntimeWarning: invalid value encountered in true_divide
  return self.params / self.bse
/Users/sergi/anaconda/lib/python3.6/site-packages/scipy/stats/_distn_infrastructure.py:879: RuntimeWarning: invalid value encountered in greater
  return (self.a < x) & (x < self.b)
/Users/sergi/anaconda/lib/python3.6/site-packages/scipy/stats/_distn_infrastructure.py:879: RuntimeWarning: invalid value encountered in less
  return (self.a < x) & (x < self.b)
/Users/sergi/anaconda/lib/python3.6/site-packages/scipy/stats/_distn_infrastructure.py:1818: RuntimeWarning: invalid value encountered in less_equal
  cond2 = cond0 & (x <= self.a)


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sales   R-squared:                       0.981
Model:                            OLS   Adj. R-squared:                  0.981
Method:                 Least Squares   F-statistic:                     5206.
Date:                Sun, 03 Dec 2017   Prob (F-statistic):          6.73e-172
Time:                        22:59:02   Log-Likelihood:                -426.71
No. Observations:                 200   AIC:                             857.4
Df Residuals:                     198   BIC:                             864.0
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const               0          0        nan        nan           0           0
TV             0.0548      0.001     42.962      0.000       0.052       0.057
radio          0.2356      0.008     29.909      0.000       0.220       0.251
==============================================================================
Omnibus:                        6.047   Durbin-Watson:                   2.080
Prob(Omnibus):                  0.049   Jarque-Bera (JB):                8.829
Skew:                          -0.112   Prob(JB):                       0.0121
Kurtosis:                       4.005   Cond. No.                          inf
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is      0. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

## 2.4. Regression model with multiple variables and without a constant for Situation 2

> Put your code here

In [27]:
X = df[["TV", "radio", "newspaper"]]
y = df["sales"]

In [28]:
X = sm.add_constant(X) 
X.head()

const     TV  radio  newspaper
1    1.0  230.1   37.8       69.2
2    1.0   44.5   39.3       45.1
3    1.0   17.2   45.9       69.3
4    1.0  151.5   41.3       58.5
5    1.0  180.8   10.8       58.4

In [29]:
X["const"] = X["const"].apply(lambda x: x*0)
X

const     TV  radio  newspaper
1      0.0  230.1   37.8       69.2
2      0.0   44.5   39.3       45.1
3      0.0   17.2   45.9       69.3
4      0.0  151.5   41.3       58.5
5      0.0  180.8   10.8       58.4
6      0.0    8.7   48.9       75.0
7      0.0   57.5   32.8       23.5
8      0.0  120.2   19.6       11.6
9      0.0    8.6    2.1        1.0
10     0.0  199.8    2.6       21.2
11     0.0   66.1    5.8       24.2
12     0.0  214.7   24.0        4.0
13     0.0   23.8   35.1       65.9
14     0.0   97.5    7.6        7.2
15     0.0  204.1   32.9       46.0
16     0.0  195.4   47.7       52.9
17     0.0   67.8   36.6      114.0
18     0.0  281.4   39.6       55.8
19     0.0   69.2   20.5       18.3
20     0.0  147.3   23.9       19.1
21     0.0  218.4   27.7       53.4
22     0.0  237.4    5.1       23.5
23     0.0   13.2   15.9       49.6
24     0.0  228.3   16.9       26.2
25     0.0   62.3   12.6       18.3
26     0.0  262.9    3.5       19.5
27     0.0  142.9   29.3       12.6
28     0.0  240.1   16.7       22.9
29     0.0  248.8   27.1       22.9
30     0.0   70.6   16.0       40.8
..     ...    ...    ...        ...
171    0.0   50.0   11.6       18.4
172    0.0  164.5   20.9       47.4
173    0.0   19.6   20.1       17.0
174    0.0  168.4    7.1       12.8
175    0.0  222.4    3.4       13.1
176    0.0  276.9   48.9       41.8
177    0.0  248.4   30.2       20.3
178    0.0  170.2    7.8       35.2
179    0.0  276.7    2.3       23.7
180    0.0  165.6   10.0       17.6
181    0.0  156.6    2.6        8.3
182    0.0  218.5    5.4       27.4
183    0.0   56.2    5.7       29.7
184    0.0  287.6   43.0       71.8
185    0.0  253.8   21.3       30.0
186    0.0  205.0   45.1       19.6
187    0.0  139.5    2.1       26.6
188    0.0  191.1   28.7       18.2
189    0.0  286.0   13.9        3.7
190    0.0   18.7   12.1       23.4
191    0.0   39.5   41.1        5.8
192    0.0   75.5   10.8        6.0
193    0.0   17.2    4.1       31.6
194    0.0  166.8   42.0        3.6
195    0.0  149.7   35.6        6.0
196    0.0   38.2    3.7       13.8
197    0.0   94.2    4.9        8.1
198    0.0  177.0    9.3        6.4
199    0.0  283.6   42.0       66.2
200    0.0  232.1    8.6        8.7

[200 rows x 4 columns]

In [30]:
model = sm.OLS(y, X).fit() ## sm.OLS(output, input)

In [31]:
predictions = model.predict(X) 
predictions

1      21.941397
2      11.885691
3      12.290830
4      18.311216
5      13.107703
6      12.596134
7      10.777274
8      11.016509
9       0.946104
10     11.681915
11      5.251518
12     16.949837
13     10.188625
14      7.054714
15     19.063754
16     22.000762
17     13.697677
18     24.875590
19      8.585798
20     13.555969
21     18.801838
22     14.298736
23      5.077560
24     16.476916
25      6.459038
26     15.247600
27     14.410006
28     17.011721
29     19.790874
30      8.039452
         ...    
171     5.576852
172    14.290411
173     5.806968
174    10.851616
175    12.939181
176    26.464811
177    20.414540
178    11.480687
179    15.793884
180    11.426176
181     9.141175
182    13.414322
183     4.789246
184    26.233735
185    18.890312
186    21.379388
187     8.417960
188    16.963610
189    18.535657
190     4.088362
191    11.355857
192     6.562240
193     2.367750
194    18.366571
195    16.064835
196     3.109157
197     6.292321
198    11.6955

In [32]:
model.summary()

/Users/sergi/anaconda/lib/python3.6/site-packages/statsmodels/regression/linear_model.py:1471: RuntimeWarning: divide by zero encountered in double_scalars
  return np.sqrt(eigvals[0]/eigvals[-1])
/Users/sergi/anaconda/lib/python3.6/site-packages/statsmodels/base/model.py:1036: RuntimeWarning: invalid value encountered in true_divide
  return self.params / self.bse
/Users/sergi/anaconda/lib/python3.6/site-packages/scipy/stats/_distn_infrastructure.py:879: RuntimeWarning: invalid value encountered in greater
  return (self.a < x) & (x < self.b)
/Users/sergi/anaconda/lib/python3.6/site-packages/scipy/stats/_distn_infrastructure.py:879: RuntimeWarning: invalid value encountered in less
  return (self.a < x) & (x < self.b)
/Users/sergi/anaconda/lib/python3.6/site-packages/scipy/stats/_distn_infrastructure.py:1818: RuntimeWarning: invalid value encountered in less_equal
  cond2 = cond0 & (x <= self.a)


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sales   R-squared:                       0.982
Model:                            OLS   Adj. R-squared:                  0.982
Method:                 Least Squares   F-statistic:                     3566.
Date:                Sun, 03 Dec 2017   Prob (F-statistic):          2.43e-171
Time:                        22:59:10   Log-Likelihood:                -423.54
No. Observations:                 200   AIC:                             853.1
Df Residuals:                     197   BIC:                             863.0
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const               0          0        nan        nan           0           0
TV             0.0538      0.001     40.507      0.000       0.051       0.056
radio          0.2222      0.009     23.595      0.000       0.204       0.241
newspaper      0.0168      0.007      2.517      0.013       0.004       0.030
==============================================================================
Omnibus:                        5.982   Durbin-Watson:                   2.038
Prob(Omnibus):                  0.050   Jarque-Bera (JB):                7.039
Skew:                          -0.232   Prob(JB):                       0.0296
Kurtosis:                       3.794   Cond. No.                          inf
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is      0. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

### Interpreting the fitness of the models

> Answer question here

Regression model with multiple variables (TV, radio and newspaper) and with a constant for Situation 1

R squared= 0.897 ; P value: Extremely low.



Regression model with multiple variables (TV, radio and newspaper) and with a constant for Situation 2

R squared= 0.897 ; P value: Extremely low except for newspaper that is higher 0.860)



Regression model with multiple variables (TV and radio) and without a constant for Situation 1

R squared= 0.981 ; P value: Extremely low.



Regression model with multiple variables (TV, radio and newspaper) and with a constant for Situation 2

R squared= 0.982 ; P value: Extremely low except for newspaper that is higher 0.013)



We can see that the model fits very well with a constant, but fits so much better without a constant.

When we add the 3rd feature of newspaper, the model does not improve with a constant and we obtain a p-value for the newspaper values really high.

In the model without a constant, the improvement of the feature newspaper is so little (0.001 R squared) and has a significant p-value.

So, the newspaper feature is not necessary at all to build our model.